### PR TITLE
Support CnosDB

### PR DESCRIPTION
**Pull request by ZuoTiJia**
Hello, I am from the CnosDB development team. We implemented Sqlancer of CnosDB referring to the code of Postgres in Sqlancer. This tool helped us find some bugs.

### CODE REVIEW

**Comment by mrigger** [comment]
Awesome, thanks for the PR! Starting to review the PR asap. In the meanwhile, could you please execute `mvn -B verify -DskipTests=true` and fix the warnings?

**Comment by ZuoTiJia** [comment]
> Awesome, thanks for the PR! Starting to review the PR asap. In the meanwhile, could you please execute `mvn -B verify -DskipTests=true` and fix the warnings?

Thank you for reminding

File: src/sqlancer/cnosdb/CnosDBComparatorHelper.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+import sqlancer.cnosdb.client.CnosDBResultSet;
+import sqlancer.cnosdb.query.CnosDBSelectQuery;
+import sqlancer.common.query.ExpectedErrors;
+
+public final class CnosDBComparatorHelper {
+
+    private CnosDBComparatorHelper() {
+    }
+
+    public static boolean isEqualDouble(String first, String second) {
```
Comment (mrigger) [inline src/sqlancer/cnosdb/CnosDBComparatorHelper.java:21]: For this and other methods, is there actually a need to re-implement them? It seems that, at least for this method and the one below, the code is the same as for https://github.com/sqlancer/sqlancer/blob/67751dc3d83b880cd8c7ed53416b8b157a2ba2ea/src/sqlancer/ComparatorHelper.java#L21-L37

File: src/sqlancer/cnosdb/CnosDBGlobalState.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    public static final char IMMUTABLE = 'i';
+    public static final char STABLE = 's';
+    public static final char VOLATILE = 'v';
+
+    private final Map<String, Character> functionsAndTypes = new HashMap<>();
+    private final List<Character> allowedFunctionTypes = Arrays.asList(IMMUTABLE, STABLE, VOLATILE);
+
+    @Override
+    protected void executeEpilogue(Query<?> q, boolean success, ExecutionTimer timer) throws Exception {
+        boolean logExecutionTime = getOptions().logExecutionTime();
```
Comment (mrigger) [inline src/sqlancer/cnosdb/CnosDBGlobalState.java:24]: This seems to be a copy of https://github.com/sqlancer/sqlancer/blob/67751dc3d83b880cd8c7ed53416b8b157a2ba2ea/src/sqlancer/SQLGlobalState.java#L19-L27

I think it should be possible to inherit from `SQLGlobalState` instead?

File: src/sqlancer/cnosdb/CnosDBLoggableFactory.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        errors.addAll(CnosDBExpectedError.expectedErrors());
+        return new CnosDBOtherQuery(queryString, errors);
+    }
+
+    @Override
+    public CnosDBQueryAdapter commentOutQuery(Query<?> query) {
+        String queryString = query.getLogString();
+        String newQueryString = "-- " + queryString;
+        ExpectedErrors errors = new ExpectedErrors();
+        errors.addAll(CnosDBExpectedError.expectedErrors());
```
Comment (mrigger) [inline src/sqlancer/cnosdb/CnosDBLoggableFactory.java:40]: I believe that commenting out a SQL statement should not result in any errors, so the expected errors could be empty here?

File: src/sqlancer/cnosdb/CnosDBLoggableFactory.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+import sqlancer.cnosdb.query.CnosDBOtherQuery;
+import sqlancer.cnosdb.query.CnosDBQueryAdapter;
+import sqlancer.common.log.Loggable;
+import sqlancer.common.log.LoggableFactory;
+import sqlancer.common.log.LoggedString;
+import sqlancer.common.query.ExpectedErrors;
+import sqlancer.common.query.Query;
+
+public class CnosDBLoggableFactory extends LoggableFactory {
```
Comment (mrigger) [inline src/sqlancer/cnosdb/CnosDBLoggableFactory.java:14]: I think inheriting from https://github.com/sqlancer/sqlancer/blob/67751dc3d83b880cd8c7ed53416b8b157a2ba2ea/src/sqlancer/common/log/SQLLoggableFactory.java#L9 and overriding methods whose implementation differs would prevent some code duplication here.

File: src/sqlancer/cnosdb/ast/CnosDBAggregate.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                CnosDBDataType.UINT),
+        COUNT(CnosDBDataType.INT) {
+            @Override
+            public CnosDBDataType[] getInputTypes(CnosDBDataType returnType) {
+                return new CnosDBDataType[] { CnosDBDataType.getRandomType() };
+            }
+        },
+        SUM(CnosDBDataType.INT, CnosDBDataType.DOUBLE, CnosDBDataType.UINT), APPROX_MEDIAN(CnosDBDataType.DOUBLE);
+
+        // Currently these aggregate functions have bugs https://github.com/cnosdb/cnosdb/issues/786
```
Comment (mrigger) [inline src/sqlancer/cnosdb/ast/CnosDBAggregate.java:33]: As an optional suggestion, I think it would be more elegant to include the currently commented-out code. If, for example, you require a random aggregate function, you could still remove the buggy aggregate functions from the selection. The general pattern is described at https://github.com/sqlancer/sqlancer/blob/67751dc3d83b880cd8c7ed53416b8b157a2ba2ea/CONTRIBUTING.md#implementing-support-for-a-new-dbms, and I see you added a `CnosDBBugs` class already.

File: src/sqlancer/cnosdb/ast/CnosDBBinaryArithmeticOperation.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+public class CnosDBBinaryArithmeticOperation extends BinaryOperatorNode<CnosDBExpression, CnosDBBinaryOperator>
+        implements CnosDBExpression {
+
+    public CnosDBBinaryArithmeticOperation(CnosDBExpression left, CnosDBExpression right, CnosDBBinaryOperator op) {
+        super(left, right, op);
+    }
+
+    @Override
+    public CnosDBConstant getExpectedValue() {
```
Comment (mrigger) [inline src/sqlancer/cnosdb/ast/CnosDBBinaryArithmeticOperation.java:21]: Currently, the PR includes many `getExpectedValue` implementations, but does not support PQS, for which they are needed. Perhaps we can remove these implementations?

File: src/sqlancer/cnosdb/ast/CnosDBBinaryArithmeticOperation.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+        public static CnosDBBinaryOperator getRandom(CnosDBDataType dataType) {
+            List<CnosDBBinaryOperator> ops = new ArrayList<>(Arrays.asList(values()));
+            switch (dataType) {
+            case DOUBLE:
+            case UINT:
+            case STRING:
+                ops.remove(MODULO);
+                ops.remove(EXPONENTIATION);
+                ops.remove(EXPONENTIATION);
```
Comment (mrigger) [inline src/sqlancer/cnosdb/ast/CnosDBBinaryArithmeticOperation.java:104]: Why are these removed twice? A comment to explain this would also be helpful.

File: src/sqlancer/cnosdb/ast/CnosDBFunction.java [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+                return type == CnosDBDataType.STRING;
+            }
+
+            @Override
+            public CnosDBDataType[] getInputTypesForReturnType(CnosDBDataType returnType, int nrArguments) {
+                return new CnosDBDataType[] { CnosDBDataType.STRING };
+            }
+
+        };
+        // NULL_IF(2, "nullif") {
```
Comment (mrigger) [inline src/sqlancer/cnosdb/ast/CnosDBFunction.java:140]: Perhaps we can remove this commented-out code?

File: src/sqlancer/cnosdb/ast/CnosDBFunctionWithUnknownResult.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    ACOS("acos", CnosDBDataType.DOUBLE, CnosDBDataType.DOUBLE),
+    ASIN("asin", CnosDBDataType.DOUBLE, CnosDBDataType.DOUBLE),
+    ATAN2("atan2", CnosDBDataType.DOUBLE, CnosDBDataType.DOUBLE, CnosDBDataType.DOUBLE),
+    COS("cos", CnosDBDataType.DOUBLE, CnosDBDataType.DOUBLE), SIN("sin", CnosDBDataType.DOUBLE, CnosDBDataType.DOUBLE),
+    SQRT("sqrt", CnosDBDataType.DOUBLE, CnosDBDataType.DOUBLE),
+    TAN("tan", CnosDBDataType.DOUBLE, CnosDBDataType.DOUBLE),
+    DATA_PART("date_part", CnosDBDataType.INT, CnosDBDataType.STRING, CnosDBDataType.TIMESTAMP);
+
+    // because bug of arrow-csv https://github.com/apache/arrow-rs/issues/3547
+    // NOW("now", CnosDBDataType.TIMESTAMP),
```
Comment (mrigger) [inline src/sqlancer/cnosdb/ast/CnosDBFunctionWithUnknownResult.java:66]: As an optional suggestion, as in the other instance, we could apply the `CnosDBBugs` pattern here.

**Review by mrigger — COMMENTED** [review]
Thanks a lot for the PR! Could you please take a look at my comments, which are minor issues and suggestions?

Currently, the implementation contains much effectively dead code, namely the `getExpectedValue` methods needed for Pivoted Query Synthesis (PQS), which is not supported as a test oracle. Would it be possible to remove these methods?

File: src/sqlancer/cnosdb/CnosDBGlobalState.java (hunk 2 above)
Comment (ZuoTiJia) [reply src/sqlancer/cnosdb/CnosDBGlobalState.java:12]: SQLGlobalState need SQLConnection. Currently our implementation does not use jdbc.

**Comment by ZuoTiJia** [comment]
Thanks for your review, I removed some dead code.

File: src/sqlancer/cnosdb/CnosDBLoggableFactory.java (hunk 4 above)
Comment (mrigger) [reply src/sqlancer/cnosdb/CnosDBLoggableFactory.java:14]: I noticed that this is still unaddressed. Is it not possible to inherit because of some reason related to not using JDBC?

File: src/sqlancer/cnosdb/CnosDBGlobalState.java (hunk 2 above)
Comment (mrigger) [reply src/sqlancer/cnosdb/CnosDBGlobalState.java:12]: Okay, I see. It would be ideal to still re-use some code (e.g., by factoring out methods), but this could also be as part of a follow-up PR.

File: src/sqlancer/cnosdb/CnosDBExpectedError.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class CnosDBExpectedError {
+    private CnosDBExpectedError() {
+    }
+
+    public static List<String> expectedErrors() {
+        List<String> errors = new ArrayList<>();
```
Comment (mrigger) [inline src/sqlancer/cnosdb/CnosDBExpectedError.java:11]: Here, it would be ideal to directly create an `ExpectedError` object rather than a list.

**Review by mrigger — COMMENTED** [review]
Almost ready to merge! Could you please still have a look at the other comments?

File: src/sqlancer/cnosdb/CnosDBLoggableFactory.java (hunk 4 above)
Comment (ZuoTiJia) [reply src/sqlancer/cnosdb/CnosDBLoggableFactory.java:14]: At present, our implementation does not use JDBC. At the beginning of development, JDBC support was not perfect, so we used the http interface, and we will optimize it later.

File: src/sqlancer/cnosdb/ast/CnosDBAggregate.java (hunk 5 above)
Comment (ZuoTiJia) [reply src/sqlancer/cnosdb/ast/CnosDBAggregate.java]: Thanks. I removed the comment and applied that pattern.

File: src/sqlancer/cnosdb/ast/CnosDBBinaryArithmeticOperation.java (hunk 6 above)
Comment (ZuoTiJia) [reply src/sqlancer/cnosdb/ast/CnosDBBinaryArithmeticOperation.java]: Thanks, I remove those dead code.

File: src/sqlancer/cnosdb/ast/CnosDBFunction.java (hunk 8 above)
Comment (ZuoTiJia) [reply src/sqlancer/cnosdb/ast/CnosDBFunction.java]: Thanks, I remove those comment.

File: src/sqlancer/cnosdb/ast/CnosDBFunctionWithUnknownResult.java (hunk 9 above)
Comment (ZuoTiJia) [reply src/sqlancer/cnosdb/ast/CnosDBFunctionWithUnknownResult.java]: Thanks, I remove those comment.

File: src/sqlancer/cnosdb/CnosDBComparatorHelper.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+            String originalQueryString, List<String> combinedString, CnosDBGlobalState state) {
+        if (resultSet.size() != secondResultSet.size()) {
+            String queryFormatString = "-- %s;\n-- cardinality: %d";
+            String firstQueryString = String.format(queryFormatString, originalQueryString, resultSet.size());
+            String secondQueryString = String.format(queryFormatString, String.join(";", combinedString),
+                    secondResultSet.size());
+            state.getState().getLocalState().log(String.format("%s\n%s", firstQueryString, secondQueryString));
+            String assertionMessage = String.format("the size of the result sets mismatch (%d and %d)!\n%s\n%s",
+                    resultSet.size(), secondResultSet.size(), firstQueryString, secondQueryString);
+            Main.nrUnsuccessfulActions.addAndGet(1);
```
Comment (mrigger) [inline src/sqlancer/cnosdb/CnosDBComparatorHelper.java:75]: The most recent commit added a couple of `Main.nrUnsuccessfulActions.addAndGet(1);` and similar calls like this. I believe we should remove them to not let the two comparator helper classes diverge until we fix the code duplication problem.

File: src/sqlancer/cnosdb/CnosDBExpectedError.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        ERRORS.add("The percentile argument for ApproxPercentileCont must be Float64, not Null.");
+        ERRORS.add("This feature is not implemented");
+        ERRORS.add("The function Avg does not support inputs of type Null.");
+        ERRORS.add("Coercion from [Utf8, Timestamp(Nanosecond, Some(\\\"+00:00\\\"))]");
+        ERRORS.add(
+                "Coercion from [Utf8, Float64, Utf8] to the signature OneOf([Exact([Utf8, Int64]), Exact([LargeUtf8, Int64]), Exact([Utf8, Int64, Utf8]), Exact([LargeUtf8, Int64, Utf8]), Exact([Utf8, Int64, LargeUtf8]), Exact([LargeUtf8, Int64, LargeUtf8])]) failed.");
+        ERRORS.add("Coercion from");
+    }
+
+    public static ExpectedErrors expectedErrors() {
```
Comment (mrigger) [inline src/sqlancer/cnosdb/CnosDBExpectedError.java:74]: I think that this method should be called `getExpectedExpressionErrors` since all the errors seem specific to expressions? This is something that could be done in a future PR.

**Review by mrigger — APPROVED** [review]
LGTM, thanks a lot again for your contribution! It would be great to add CI tests (e.g., see https://github.com/sqlancer/sqlancer/blob/39c48cde9ff0ab29b950a8d3e35930b0949f36a0/.github/workflows/main.yml#L105) as one of the next steps, so we make sure that the implementation does not break in the future.

**Comment by mrigger** [comment]
The branch is currently out-of-date. Could please merge latest the latest master? It seems I don't have permissions to do so.

**Comment by ZuoTiJia** [comment]
> LGTM, thanks a lot again for your contribution! It would be great to add CI tests (e.g., see
> 
> https://github.com/sqlancer/sqlancer/blob/39c48cde9ff0ab29b950a8d3e35930b0949f36a0/.github/workflows/main.yml#L105
> 
> ) as one of the next steps, so we make sure that the implementation does not break in the future.

we'll add it soon

**Comment by mrigger** [comment]
Thanks a lot!
